src/ROOTNTupleWriter.cc

```diff
@@ -0,0 +1,233 @@
+  if (!m_finished) {
+    finish();
+  }
+}
+
+template <typename T>
+std::pair<std::vector<std::string>&, std::vector<std::vector<T>>&>
+ROOTNTupleWriter::getKeyValueVectors() {
+  if constexpr (std::is_same_v<T, int>) {
+    return {m_intkeys, m_intvalues};
+  } else if constexpr (std::is_same_v<T, float>) {
+    return {m_floatkeys, m_floatvalues};
+  } else if constexpr (std::is_same_v<T, double>) {
+    return {m_doublekeys, m_doublevalues};
+  } else if constexpr (std::is_same_v<T, std::string>) {
+    return {m_stringkeys, m_stringvalues};
+  } else {
+    throw std::runtime_error("Unknown type");
+  }
+}
+
+template <typename T>
+void ROOTNTupleWriter::fillParams(GenericParameters& params, ROOT::Experimental::REntry *entry) {
+  auto [key, value] = getKeyValueVectors<T>();
+  entry->CaptureValueUnsafe(root_utils::getGPKeyName<T>(), &key);
+  entry->CaptureValueUnsafe(root_utils::getGPValueName<T>(), &value);
+
+  key.clear();
+  key.reserve(params.getMap<T>().size());
+  value.clear();
+  value.reserve(params.getMap<T>().size());
+
+  for (auto& [kk, vv] : params.getMap<T>()) {
+    key.emplace_back(kk);
+    value.emplace_back(vv);
+  }
+
+}
+
+void ROOTNTupleWriter::writeFrame(const podio::Frame& frame, const std::string& category) {
+  writeFrame(frame, category, frame.getAvailableCollections());
+}
+
+void ROOTNTupleWriter::writeFrame(const podio::Frame& frame, const std::string& category,
+                                  const std::vector<std::string>& collsToWrite) {
+
+  std::vector<StoreCollection> collections;
+  collections.reserve(collsToWrite.size());
+  for (const auto& name : collsToWrite) {
+    auto* coll = frame.getCollectionForWrite(name);
+    collections.emplace_back(name, const_cast<podio::CollectionBase*>(coll));
+  }
+
+  bool new_category = false;
+  if (m_writers.find(category) == m_writers.end()) {
+    new_category = true;
+    auto model = createModels(collections);
+    m_writers[category] = ROOT::Experimental::RNTupleWriter::Append(std::move(model), category, *m_file.get(), {});
+  }
+
+  auto entry = m_writers[category]->GetModel()->CreateBareEntry();
+
+  ROOT::Experimental::RNTupleWriteOptions options;
+  options.SetCompression(ROOT::RCompressionSetting::EDefaults::kUseGeneralPurpose);
+
+  for (const auto& [name, coll] : collections) {
+    auto collBuffers = coll->getBuffers();
+    if (collBuffers.vecPtr) {
+      entry->CaptureValueUnsafe(name, (void*)collBuffers.vecPtr);
+    }
+
+    if (auto refColls = collBuffers.references) {
+      int i = 0;
+      for (auto& c : (*refColls)) {
+        const auto brName = root_utils::refBranch(name, i++);
+        entry->CaptureValueUnsafe(brName, c.get());
+      }
+    }
+
+    if (auto vmInfo = collBuffers.vectorMembers) {
+      int i = 0;
+      for (auto& [type, vec] : (*vmInfo)) {
+        const auto typeName = "vector<" + type + ">";
+        const auto brName = root_utils::vecBranch(name, i++);
+        auto ptr = *(std::vector<int>**)vec;
```

Comment:
```suggestion
        const auto brName = root_utils::vecBranch(name, i++);
        auto ptr = *(std::vector<int>**)vec;
```
The `typeName` seems to be unused, so I suppose this works because the necessary type information has been passed to the whole thing when we create the fields? And the `vector<int>` cast works for all types that potentially come along here and is simply necessary for getting to the correct `void*`?

Reply:
Yeah I don't think there is a typename for the other members, must have forgotten to remove it when copying and pasting from the TTree reader. I should double check that it works for every type, but in principle it's the same three star cast used in other places: https://github.com/AIDASoft/podio/blob/930e6001e893bef9415639710e12b67b5deba5d6/include/podio/CollectionBuffers.h#L41